bigquery/storage_integration_test.go

```diff
@@ -418,11 +423,82 @@ func TestIntegration_StorageReadCancel(t *testing.T) {
 	}
 	// resources are cleaned asynchronously
 	time.Sleep(time.Second)
-	if !it.arrowIterator.isDone() {
+	arrowIt := it.arrowIterator.(*storageArrowIterator)
+	if !arrowIt.isDone() {
 		t.Fatal("expected stream to be done")
 	}
 }
 
+func TestIntegration_StorageReadArrow(t *testing.T) {
+	if client == nil {
+		t.Skip("Integration tests skipped")
+	}
+	ctx := context.Background()
+	table := "`bigquery-public-data.usa_names.usa_1910_current`"
+	sql := fmt.Sprintf(`SELECT name, number, state FROM %s where state = "CA"`, table)
+
+	q := storageOptimizedClient.Query(sql)
+	job, err := q.Run(ctx) // force usage of Storage API by skipping fast paths
+	if err != nil {
+		t.Fatal(err)
+	}
+	it, err := job.Read(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	arrowIt, err := it.ArrowIterator()
+	if err != nil {
+		t.Fatalf("expected iterator to be accelerated: %v", err)
+	}
+	arrowItReader := NewArrowIteratorReader(arrowIt)
+
+	records := []arrow.Record{}
+	r, err := ipc.NewReader(arrowItReader)
+	numrec := 0
+	for r.Next() {
+		rec := r.Record()
+		rec.Retain()
```

Comment:
I'm confused by the release/retains here, but I've not been spending much time with arrow recently.  If you retain individual records do you need to release them?

Reply:
they are going to be releases by the ipc.Reader later on the `r.Release()` call right after.

Reply:
not exactly. If you call `Retain` on an individual record, then you will need to call `Release` on that record.

The `ipc.Reader` keeps only the current Record, reusing that member. When you call `Next()` it will release the record it had before loading the next one. This is why you need to call `Retain` on the records that you put into the slice, so that they aren't deallocated by the `ipc.Reader` calling `Release` on them. However you should also add a `defer rec.Release()` in the loop to ensure that record gets released, the `ipc.Reader` will not retain any references to those records and therefore will not call `Release` on them.

Reply:
interesting, I didn't know that. I'll make the changes to call `rec.Release` on each record.

Reply:
If you want to verify that everything is released/retained correctly,  you could use [`memory.CheckedAllocator`](https://pkg.go.dev/github.com/apache/arrow/go/v13@v13.0.0/arrow/memory#NewCheckedAllocator) and `defer mem.AssertSize(t, 0)` then pass the checked allocator to everything (like to `ipc.NewReader`) so that it is used for all the memory allocations. 

Not absolutely necessary, but an optional way to add some assurances if desired.

Reply:
I'm gonna add support for changing the allocator just internally for now for test purposes, I liked the idea of verifying that there are no memory leaks. Thanks for the tip.